### PR TITLE
Recovery Mode for StablePool

### LINKED ISSUE #1353: Recovery Mode (Stable Pool)
This propagates the Recovery Mode mechanism to the StablePool. It extends the basic infrastructure behavior by maintaining the _lastInvariant without needing to run through the normal invariant calculation (which is iterative, and can fail).

### PR DESCRIPTION
**Pull request by EndymionJkb**
Use recovery mode in StablePool; update the last invariant on exit.

Closes #1353

### CODE REVIEW

File: pkg/pool-stable/test/StablePool.test.ts [hunk 1]
```diff
@@ -999,5 +1150,164 @@ describe('StablePool', function () {
...
+        context('when not in recovery mode', () => {
+          it('is not in recovery mode', async () => {
+            expect(await pool.inRecoveryMode()).to.be.false;
+          });
+
+          it('recovery mode exit is not available', async () => {
+            await expect(pool.recoveryModeExit({ from: lp, bptIn: fp(1) })).to.be.revertedWith('NOT_IN_RECOVERY_MODE');
+          });
+        });
+      });
```
Comment (nventuro) [inline pkg/pool-stable/test/StablePool.test.ts:1310]: I think all of this is also out of scope, as this is BasePool behavior. We're not introducing any StablePool specific changes here, so there's nothing to test. The only recovery mode thing we should look at is that hte last invariant is updated at the end.

File: pkg/pool-stable/test/StablePool.test.ts [hunk 2]
```diff
@@ -294,13 +308,22 @@ describe('StablePool', function () {
...
             expect(result.bptOut).to.be.equalWithError(expectedBptOut, 0.0001);
+          }
+
+          it('can tell how much BPT it will give in return', async () => {
+            await joinGivenIn();
+          });
+
+          it('can tell how much BPT it will give in return (in recovery mode)', async () => {
+            await enterRecoveryMode(pool);
+            await joinGivenIn();
```
[the remark this answers is not in the document]
Comment (nventuro) [reply pkg/pool-stable/test/StablePool.test.ts:326]: This is, however, out of scope. BasePool allows for regular joins and exits outside of recovery mode, there's nothing to test here.

File: pkg/pool-stable/test/StablePool.test.ts [hunk 3]
```diff
@@ -694,21 +799,33 @@ describe('StablePool', function () {
...
               currentBalances,
               protocolFeePercentage,
             });
 
-            expect(result.dueProtocolFeeAmounts).to.be.equalWithError(expectedDueProtocolFeeAmounts, 0.0001);
+            if (await pool.inRecoveryMode()) {
+              expect(result.dueProtocolFeeAmounts).to.be.zeros;
+            } else {
+              expect(result.dueProtocolFeeAmounts).to.be.equalWithError(expectedDueProtocolFeeAmounts, 0.0001);
+            }
```
Comment (nventuro) [inline pkg/pool-stable/test/StablePool.test.ts:825]: This is also out of scope for testing, as BasePool will simply report a fee % of 0.

File: pkg/pool-utils/test/LegacyBasePool.test.ts [hunk 4]
```diff
@@ -814,4 +816,79 @@ describe('LegacyBasePool', function () {
...
+        }
+
+        // Token balances should be restored to the owner
+        tokenBalances = await Promise.all(tokens.map(async (token) => await token.balanceOf(poolOwner)));
+        for (let i = 0; i < tokenBalances.length; i++) {
+          expect(tokenBalances[i]).to.equalWithError(initialBalances[i], 0.000001);
+        }
+      });
+    });
+  });
```
Comment (nventuro) [inline pkg/pool-utils/test/LegacyBasePool.test.ts:893]: These are duplicates and should be removed - we already have tests for the exit mode (see line 594).

If we add the recovery mode protocol fee override however, we should add tests there to check that the `protocolFeePercentage` value is 0 while in recovery mode, even if the global fee is not 0. But again, ideally different PR as this is already quite busy.

File: pkg/pool-utils/test/RecoveryModePool.test.ts [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+import { actionId } from '@balancer-labs/v2-helpers/src/models/misc/actions';
+import { BigNumberish, fp } from '@balancer-labs/v2-helpers/src/numbers';
+import { PoolSpecialization } from '@balancer-labs/balancer-js';
+import { MONTH } from '@balancer-labs/v2-helpers/src/time';
+import { ANY_ADDRESS, ZERO_ADDRESS } from '@balancer-labs/v2-helpers/src/constants';
+
+import TokenList from '@balancer-labs/v2-helpers/src/models/tokens/TokenList';
+import TypesConverter from '@balancer-labs/v2-helpers/src/models/types/TypesConverter';
+
+describe('RecoveryModePool', function () {
```
Comment (nventuro) [inline pkg/pool-utils/test/RecoveryModePool.test.ts:17]: All of these are duplicates, is there a reason to add them or just a merge artifact?

File: pkg/pool-stable/test/StablePool.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (EndymionJkb) [reply pkg/pool-stable/test/StablePool.test.ts:326]: I didn't add these tests - it was like that before. I just added the "in recovery mode" rerun of it.

File: pkg/pool-stable/test/StablePool.test.ts [hunk 6]
```diff
@@ -999,5 +1150,164 @@ describe('StablePool', function () {
...
+          });
+
+          it('is in the middle of an AMP change', async () => {
+            const { value, isUpdating, precision } = await pool.getAmplificationParameter();
+            expect(value).to.be.lt(AMPLIFICATION_PARAMETER.mul(precision));
+            expect(isUpdating).to.be.true;
+
+            await expect(pool.enterRecoveryMode(admin)).to.not.be.reverted;
+          });
+        });
```
[the remark this answers is not in the document]
Comment (EndymionJkb) [reply pkg/pool-stable/test/StablePool.test.ts:1202]: Ok; it was relevant when we were blocking amp changes.

File: pkg/pool-utils/contracts/LegacyBasePool.sol [hunk 7]
```diff
@@ -337,7 +337,7 @@ abstract contract LegacyBasePool is IBasePool, BalancerPoolToken, TemporarilyPau
                 recipient,
                 balances,
                 lastChangeBlock,
-                protocolSwapFeePercentage,
+                inRecoveryMode() ? 0 : protocolSwapFeePercentage,
```
[the remark this answers is not in the document]
Comment (EndymionJkb) [reply pkg/pool-utils/contracts/LegacyBasePool.sol:340]: Ok; will back out and make a new PR.

File: pkg/pool-stable/test/StablePool.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (nventuro) [reply pkg/pool-stable/test/StablePool.test.ts:330]: That's what I mean - we don't need to test that the joins and exits work properly in recovery mode because that is part of the regular behavior, there's nothing special about it. But if you want to add it for completeness, I'd do it in the way suggested above.
